iroha/src/sumeragi.rs

```diff
@@ -1,14 +1,214 @@
-use crate::prelude::*;
+use crate::{
+    crypto::{Crypto, Hash},
+    peer::{self, PeerId},
+    prelude::*,
+};
+use iroha_derive::*;
+use parity_scale_codec::{Decode, Encode};
+use std::cmp::Ordering;
 
-#[derive(Default)]
-pub struct Sumeragi {}
+pub struct Sumeragi {
+    sorted_peers: Vec<PeerId>,
+    max_faults: usize,
+    peer_id: PeerId,
+    /// Block in discussion this round
+    pending_block: Option<Block>,
+}
+
+#[derive(Io, Decode, Encode, Debug, Clone)]
+pub enum Message {
+    /// Is sent by leader to all validating peers, when a new block is created.
+    Created(Block),
+    /// Is sent by validating peers to proxy tail and observing peers when they have signed this block.
+    Signed(Block),
+    /// Is sent by proxy tail to validating peers and to leader, when the block is committed.
+    Committed(Block),
+}
+
+#[derive(Eq, PartialEq, Debug)]
+pub enum Role {
+    Leader,
+    ValidatingPeer,
+    ObservingPeer,
+    ProxyTail,
+}
 
 impl Sumeragi {
-    pub fn new() -> Self {
-        Sumeragi {}
+    pub fn new(
+        peers: &[PeerId],
+        prev_block_hash: Option<Hash>,
+        max_faults: usize,
+    ) -> Result<Self, String> {
+        if peers.len() >= 3 * max_faults + 1 {
+            let mut peers = peers.to_vec();
+            Self::sort_peers(&mut peers, prev_block_hash);
```

Comment:
Maybe we can use sorted collection?

Reply:
Ordering is determined not only by PeerId but also by previous block hash, so we still would need to resort it every new round. Taking into account that we sort all the peer only in the beginning of each round, I don't think using sorted collection would improve anything.

Reply:
https://soramitsu.atlassian.net/browse/RH2-67